### PR TITLE
Api 3839 refactor dynamo client

### PR DESCRIPTION
**Pull request by crolarlibertyva**
Dynamo client is now a class that only uses the document api for dynamo. Use of the client changed accordingly.

### CODE REVIEW

File: oauth-proxy/dynamo_client.js [hunk 1]
```diff
@@ -1,222 +1,133 @@
...
-  };
-
-  return new Promise((resolve, reject) => {
-    client.query(params, (err, data) => {
-      if (err) {
-        reject(err);
-      } else {
-        resolve(data.Items[0]);
-      }
+  savePayloadToDynamo(payload, tableName) {
```
Comment (beaugrantham-va) [inline oauth-proxy/dynamo_client.js:23]: Please add docs to all of these.

**Review by beaugrantham-va — CHANGES_REQUESTED** [review]
In general it's not clear to me when I would use `getPayloadFromDynamo` vs `queryFromDynamo`

**Comment by lukemarkwordtlibertyits** [comment]
Overall I do not have any suggestions to this branch. I like the refactoring. Having a DynamoClient that was dependent on a Dynamo Object was a strange pattern. I am glad this takes care of it. Also, the old DynamoClient naming conventions were obscured as we added more tables. For example, `saveToDynamo` saved to the refresh token table where as `saveToDynamoAccessToken` saved to the client credentials table. Its good that the methods are more generic now. Finally, this PR increased the overall code coverage from 84% to 87%!

File: oauth-proxy/README.md [hunk 2]
```diff
@@ -30,6 +30,12 @@ Start the Oauth Proxy
 npm start
 ```
 
+Generate jsdocs for Oauth Proxy
+```
+npm run-script jsdoc
```
Comment (lukemarkwordtlibertyits) [inline oauth-proxy/README.md:35]: should the out folder be in gitignore?

File: oauth-proxy/oauthHandlers/redirectHandler.js [hunk 3]
```diff
@@ -39,19 +36,38 @@ const redirectHandler = async (
...
+      {
+        state: state,
+      },
       config.dynamo_table_name
     );
-
-    const params = new URLSearchParams(req.query);
-    loginEnd.inc();
-    res.redirect(`${document.redirect_uri.S}?${params.toString()}`);
+    if (document && document.Item && document.Item.redirect_uri) {
```
Comment (lukemarkwordtlibertyits) [inline oauth-proxy/oauthHandlers/redirectHandler.js:45]: could be 
```
if(document?.Item?.redirect_uri)
```

I guess it depends on preference

File: oauth-proxy/oauthHandlers/redirectHandler.js [hunk 4]
```diff
@@ -39,19 +36,38 @@ const redirectHandler = async (
...
     logger.error("Failed to redirect to the OAuth client application", error);
-    return next(error); // This error is unrecoverable because we can't look up the original redirect.
+    // This error is unrecoverable because we can't look up the original redirect.
+    res.status(400).json({
+      error: "invalid_request",
+      error_description:
+        "Unable to look up the original redirect for the given state",
+    });
   }
+  // Only would reach here upon an error
```
Comment (lukemarkwordtlibertyits) [inline oauth-proxy/oauthHandlers/redirectHandler.js:69]: I'm not sure this is true. res.status will not return a function, with that being said, this next statement will have no outwards issues associated with it. You will probably see one of these errors in the logs

```
(node:6) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'headers' of undefined
oauth-proxy_1         |     at setProxyResponse (/home/node/index.js:96:18)
oauth-proxy_1         |     at /home/node/index.js:128:9
oauth-proxy_1         |     at processTicksAndRejections (internal/process/task_queues.js:97:5)
oauth-proxy_1         | (node:6) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 7)
oauth-proxy_1         | (node:6) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

File: oauth-proxy/README.md (hunk 2 above)
Comment (crolarlibertyva) [reply oauth-proxy/README.md:35]: yes.

File: oauth-proxy/oauthHandlers/redirectHandler.js (hunk 3 above)
Comment (crolarlibertyva) [reply oauth-proxy/oauthHandlers/redirectHandler.js:45]: I like that better.

File: oauth-proxy/oauthHandlers/redirectHandler.js (hunk 4 above)
Comment (crolarlibertyva) [reply oauth-proxy/oauthHandlers/redirectHandler.js:69]: Will check it out. This inadvertently fixed the 500 error that was referred to during grooming.

File: oauth-proxy/oauthHandlers/redirectHandler.js (hunk 3 above)
Comment (crolarlibertyva) [reply oauth-proxy/oauthHandlers/redirectHandler.js:45]: Had to revert to get it to run in docker.

File: oauth-proxy/oauthHandlers/redirectHandler.js (hunk 4 above)
Comment (crolarlibertyva) [reply oauth-proxy/oauthHandlers/redirectHandler.js:69]: Didn't see this in logs for regression test.

File: oauth-proxy/oauthHandlers/redirectHandler.js (hunk 3 above)
Comment (crolarlibertyva) [reply oauth-proxy/oauthHandlers/redirectHandler.js:45]: We should consider an upgraded docker base image.

File: oauth-proxy/dynamo_client.js [hunk 5]
```diff
@@ -1,222 +1,172 @@
...
-  };
-
-  return new Promise((resolve, reject) => {
-    dynamoDb.dbDocClient.scan(params, function (err, data) {
-      if (err) {
-        reject(err);
-      } else {
-        resolve(data);
-      }
+    params.UpdateExpression = params.UpdateExpression.slice(0, -1);
```
Comment (beaugrantham-va) [inline oauth-proxy/dynamo_client.js:158]: Should this be updated like the other to remove the arbitrary slice?

File: oauth-proxy/oauthHandlers/redirectHandler.js [hunk 6]
```diff
@@ -39,19 +36,38 @@ const redirectHandler = async (
...
+      loginEnd.inc();
+      res.redirect(`${document.redirect_uri}?${params.toString()}`);
+    } else {
+      logger.error(
+        "Failed to get the redirect for the OAuth client application"
+      );
+      res.status(400).json({
+        error: "invalid_request",
+        error_description:
+          "Unable to look up the original redirect for the given state",
```
Comment (beaugrantham-va) [inline oauth-proxy/oauthHandlers/redirectHandler.js:57]: I would opt to be less verbose here (and below). Something like "Invalid state". I feel like this message exposes application detail/logic that is unnecessary to the user.

**Review by lukemarkwordtlibertyits — APPROVED** [review]
Overall looks good. This refactor helps clean the repository quite a bit.

**Review by dbrown-va — APPROVED** [review]
Confirmed running locally and the earlier changes requested have been implemented.

**Pull request merged**